Particle Photon 2 board support

peter-ginchev opened the pull request.

peter-ginchev committed:
Add Particle Photon 2 configuration

rei-vilo commented:
Thank you for your pull request.

Please find some comments below:

* The Particle board uses `D1`..`D4` as pins names instead of numbers. This rises a blocking error when compiling against a board that uses numbers only as pins names.
* **PDLS_Common** is used by many other libraries, which are extensively tested against each of the listed boards. Adding a new board would require owning the Particle Photon 2 and performing the same test protocol against it.
* Skipping `hV_HAL_begin()` would limit the portability of the code from one board to another., as it initialises the ports.

rei-vilo commented:
Thank you for your pull request.

Please find some comments below:

* The Particle board uses `D1`..`D4` as pins names instead of numbers. This rises a blocking error when compiling against a board that uses numbers only as pins names.
* **PDLS_Common** is used by many other libraries: they are fully tested against each of the listed boards. Adding a new board would require owning the Particle Photon 2 and performing the same test protocol against it.
* Skipping `hV_HAL_begin()` may limit the portability of the code from one board to another., as it initialises the ports.

peter-ginchev commented:
> Thank you for your pull request.
> 
> Please find some comments below:
> 
> * The Particle board uses `D1`..`D4` as pins names instead of numbers. This rises a blocking error when compiling against a board that uses numbers only as pins names.
> * **PDLS_Common** is used by many other libraries, which are extensively tested against each of the listed boards. Adding a new board would require owning the Particle Photon 2 and performing the same test protocol against it.
> * Skipping `hV_HAL_begin()` would limit the portability of the code from one board to another., as it initialises the ports.

Thank you for taking the time to review my changes,
First let state my goal -- this is my fun project with IoT, first of its kind, so excuse me if I don't have the right approach or use the incorrect language. (outside of IoT I'm a SW engineer by heart and profession)

So, to your points:
D1 .. D4 - would you prefer to have the protected by a preprocessor "#ifdef" or to use numbers, they both have downside -- ifdef would make the definition hidden, except for somebody building for it, numbers means we limit Particle to not change the numbers and keep the names -- your call -- what do you think is more relevant?

Adding Photon 2 support for PDLS_Common -- I am ok with giving back to the community, if doesn't take too much of my time, so maybe send me a private email with details, if relevant.

Is there an assumption that hv_HAL_begin() is called in addition to the display.begin(), I assumed it to be some "other" initialization needed for the tests, do you have a reference minimal "begin" and "end"?
The main reason I did that is because it seems "unnatural" for Pervasive_Wide_Small lib to use hV_HAL_SPI3_... functions without initializing h_pinSPI3 fields.

rei-vilo commented:
## Code flow

Normally, `hV_HAL_begin()` calls by default https://github.com/rei-vilo/PDLS_Common/blob/1fcaaa7eb6973ae6629cf5a93edf339fc902a97d/src/hV_HAL_Peripherals.cpp#L98-L102 

This defines the GPIOs used by SPI3.

Then,

* `Screen_EPD.begin()` calls [Pervasive_Wide_Small::begin()](https://github.com/PervasiveDisplays/Pervasive_Wide_Small/blob/ea309d8e80d3afce002e78e500f8116cb70f73d5/src/Pervasive_Wide_Small.cpp#L443). 

* `Pervasive_Wide_Small::begin()` calls [Pervasive_Wide_Small::COG_getDataOTP()](https://github.com/PervasiveDisplays/Pervasive_Wide_Small/blob/ea309d8e80d3afce002e78e500f8116cb70f73d5/src/Pervasive_Wide_Small.cpp#L66) with

``` cpp
    hV_HAL_SPI_end(); // With unicity check
    hV_HAL_SPI3_begin(); // Define 3-wire SPI pins
```

## Pins naming scheme

The pins definition for the previous generation of Particle Photon used numbers. Usually, the file arduino_pins.h provides the number associated to `D2`. Maybe the Photon 2 has a very similar approach.

```
///
/// @brief Particle Photon configuration, tested
/// @note D6 also CLK/SWCLK with D7 SWD/SWDIO
///
const pins_t boardParticlePhoton =
{
    .panelBusy = 5, ///< EXT3 and EXT3.1 pin 3 Red -> D5
    .panelDC = 4, ///< EXT3 and EXT3.1 pin 4 Orange -> D4
    .panelReset = 3, ///< EXT3 and EXT3.1 pin 5 Yellow -> D3
    .flashCS = 2, ///< EXT3 and EXT3.1 pin 8 Violet -> D2
    .panelCS = 1, ///< EXT3 and EXT3.1 pin 9 Grey -> D1
    .panelCSS = NOT_CONNECTED, ///< EXT3 and EXT3.1 pin 12 Grey2 -> D6
    .flashCSS = 0, ///< EXT3 pin 20 or EXT3.1 pin 11 Black2 -> D0
    .touchInt = NOT_CONNECTED, ///< EXT3-Touch pin 3 Red
    .touchReset = NOT_CONNECTED, ///< EXT3-Touch pin 4 Orange
    .panelPower = NOT_CONNECTED, ///< Optional power circuit
    .cardCS = NOT_CONNECTED, ///< Separate SD-card board
    .cardDetect = NOT_CONNECTED, ///< Separate SD-card board
};
```

peter-ginchev commented:
I didn't want to call hV_HAL_begin() since it was doing other stuff that I didn't want.
I've removed the change around that and will call hV_HAL_SPI3_define() in my code.

Changed the constants to numbers.
Thanks!

rei-vilo commented:
Great, thank you for the commit!
